bam/writer.go

```diff
@@ -107,83 +109,72 @@ func (bw *Writer) Write(r *sam.Record) error {
 	bin.writeInt32(int32(r.TempLen))
 
 	// Write variable length data.
-	wb.Write(append([]byte(r.Name), 0))
+	bin.writeString(r.Name)
+	bin.writeByte(0)
 	writeCigarOps(&bin, r.Cigar)
-	wb.Write(doublets(r.Seq.Seq).Bytes())
+	bin.write(doublets(r.Seq.Seq).Bytes())
 	if r.Qual != nil {
-		wb.Write(r.Qual)
+		bin.write(r.Qual)
 	} else {
 		for i := 0; i < r.Seq.Length; i++ {
-			wb.WriteByte(0xff)
+			bin.writeByte(0xff)
 		}
 	}
-	wb.Write(tags)
-	if wb.err != nil {
-		return wb.err
-	}
-
+	bin.write(tags)
 	_, err := bw.bg.Write(bw.buf.Bytes())
 	return err
 }
 
 func writeCigarOps(bin *binaryWriter, co []sam.CigarOp) {
 	for _, o := range co {
 		bin.writeUint32(uint32(o))
-		if bin.w.err != nil {
-			return
-		}
 	}
-	return
 }
 
 // Close closes the writer.
 func (bw *Writer) Close() error {
 	return bw.bg.Close()
 }
 
-type errWriter struct {
-	w   *bytes.Buffer
-	err error
+func (w *binaryWriter) write(p []byte) {
+	if _, err := w.w.Write(p); err != nil {
+		panic(err)
+	}
 }
 
-func (w *errWriter) Write(p []byte) (int, error) {
-	if w.err != nil {
-		return 0, w.err
+func (w *binaryWriter) writeString(p string) {
+	if _, err := w.w.WriteString(p); err != nil {
+		panic(err)
 	}
-	var n int
-	n, w.err = w.w.Write(p)
-	return n, w.err
 }
 
-func (w *errWriter) WriteByte(b byte) error {
-	if w.err != nil {
-		return w.err
+func (w *binaryWriter) writeByte(b byte) {
+	if err := w.w.WriteByte(b); err != nil {
+		panic(err)
 	}
-	w.err = w.w.WriteByte(b)
-	return w.err
 }
 
 type binaryWriter struct {
-	w   *errWriter
+	w   *bytes.Buffer
```

Comment:
```suggestion
	// No error handling on the buffer since it
	// is documented to always return nil err.
	*bytes.Buffer
```

Then all the writes are just `Write`/`WriteByte`/`WriteString` calls on the `bytes.Buffer` and the three methods above can be deleted.